### ISSUE TITLE
Trojaned version of file '/bin/diff' detected

### ISSUE THREAD

**Issue by dimagow** [wazuh/wazuh#19346]
Hello guys!

This is the issue was opened long time ago here https://github.com/ossec/ossec-hids/issues/2020 so I believe no one from Wazuh is watching/working on it, so I'm posting this again here.

Received From: archvbox->rootcheck
Rule: 510 fired (level 7) -> "Host-based anomaly detection event (rootcheck)."
Portion of the log(s):

Trojaned version of file '/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\.h|proc\.h|/dev/[^n]|^/bin/.*sh' (Generic).

--END OF NOTIFICATION

Received From: archvbox->rootcheck
Rule: 510 fired (level 7) -> "Host-based anomaly detection event (rootcheck)."
Portion of the log(s):

Trojaned version of file '/sbin/diff' detected. Signature used: 'bash|^/bin/sh|file\.h|proc\.h|/dev/[^n]|^/bin/.*sh' (Generic).

--END OF NOTIFICATION

Received From: archvbox->rootcheck
Rule: 510 fired (level 7) -> "Host-based anomaly detection event (rootcheck)."
Portion of the log(s):

Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\.h|proc\.h|/dev/[^n]|^/bin/.*sh' (Generic).

--END OF NOTIFICATION

Received From: archvbox->rootcheck
Rule: 510 fired (level 7) -> "Host-based anomaly detection event (rootcheck)."
Portion of the log(s):

Trojaned version of file '/usr/sbin/diff' detected. Signature used: 'bash|^/bin/sh|file\.h|proc\.h|/dev/[^n]|^/bin/.*sh' (Generic).

**Comment by falkowich** [comment]
Can confirm with newly installed server / agent

agent:
```
wazuh-agent    4.5.3-1
OS: Ubuntu 22.04.3 LTS
sha1sum /usr/bin/diff » e36327a2da31d8a11378ed3d0dc1f02bd54d33ec  /usr/bin/diff
https://www.virustotal.com/gui/file/d9f444a2b3a28dd8d1361e3131e65ae4c008747d756d3fe740a911750ef7c0b0
```

server:
```
Installation: Docker SingleNode
App version: 4.5.3
App revision: 02
Install date: Oct 12, 2023 @ 21:21:32.220
```

**Comment by rgevaert** [comment]
Same here: 
5096769cf79b5471bd14df7c5773b091  /usr/bin/diff

https://www.virustotal.com/gui/file/4de429713337777f44e9ef340176c2f1818c2fcfe0204ab27277595ff97dab77/detection

**Comment by RandomUser0815** [comment]
Seems to be resolved. See update in https://github.com/ossec/ossec-hids/issues/2020

**Comment by alainivars** [comment]
still same issue on Pop!_OS 22.04 LTS with Wazuh 4.7.1

**Comment by RandomUser0815** [comment]
I can confirm, that this is still an issue under Ubuntu 22.04.3 LTS (where Wazuh server is installed), all other servers, where only the agents are running, don't show this behaviour anymore.

md5sum /usr/bin/diff
96b16c8aa33bf49160dd02001f39695c  /usr/bin/diff

md5sum /bin/diff
96b16c8aa33bf49160dd02001f39695c  /bin/diff

WAZUH_VERSION":"v4.5.3

**Comment by gael-api** [comment]
Still present with Wazuh 4.7.2 and agent on up-to-date Fedora 39

**Comment by mehdiMj-ir** [comment]
I installed Wazuh yesterday and saw **Trojaned version of file detected.** messages on panel which freaked me out.
looks like It's false positive (a cording to above messages).
I'm on **v4.7.2**

```logs
_index
	wazuh-alerts-4.x-2024.02.04
agent.id
	[015](https://siem.my.domain/app/wazuh#/agents?tab=welcome&agent=015)
agent.ip
	192.168.100.2
agent.name
	[dbs-stage2](https://siem.my.domain/app/wazuh#/agents?tab=welcome&agent=015)
data.file
	/bin/diff
data.title
	Trojaned version of file detected.
decoder.name
	rootcheck
full_log
	Trojaned version of file '/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\.h|proc\.h|/dev/[^n]|^/bin/.*sh' (Generic).
id
	1707005111.604304
input.type
	log
location
	rootcheck
manager.name
	siem
rule.description
	Host-based anomaly detection event (rootcheck).
rule.firedtimes
	1,065
rule.gdpr
	IV_35.7.d
rule.groups
	ossec, rootcheck
rule.id
	[510](https://siem.my.domain/app/wazuh#/manager/rules?tab=rules&redirectRule=510)
rule.level
	7
rule.mail
	false
rule.pci_dss
	10.6.1
timestamp
	Feb 4, 2024 @ 03:35:11.992
```

**Comment by iasdeoupxe** [comment]
Looks like there are multiple issues about the same / similar message open:
- #19000
- #13278
- #11808

Could be possible that all have the same origin and some could be closed as duplicates.

**Comment by YousefNein** [comment]
Still present in wazuh 4.8

**Comment by ilsaloving** [comment]
I just stumbled upon this myself.  I'm wondering if the rootcheck stuff needs to be removed entirely... After a quick search, the most recent rules list I could find was well over 5 years old.  Most of the other results were 10.

**Comment by keeganquinn** [comment]
As this and other issues have reported, the rootcheck and SCA features are effectively unmaintained in Wazuh.

I would hesitate to suggest that these features should be removed entirely; they *should* be maintained and updated regularly. However, given the current state of the code, production installations which intend to take seriously the reports provided by this program should probably disable these features until such time as they are updated to work correctly. Such systems should use alternative programs to provide this functionality instead.

The following settings may be used to disable these unmaintained features:

```
  <rootcheck>
    <disabled>yes</disabled>
  </rootcheck>

  <sca>
    <enabled>no</enabled>
  </sca>
```

In my opinion, these settings should probably be applied by default in the stock configuration files if there are no plans to fix the associated features in the immediate future. These types of alarming false positives damage the trustworthiness of the product as a whole.

**Comment by inf0rmix** [comment]
We also discovered this falsepositive and wonder whats the problem aout whitelisting this specific hash. Disabling the whole check is not a trustworthy solution for such a new and promising project.

**Comment by lfkdev** [comment]
Still a problem, shocked me for a minute. `Ubuntu 22 4.7.3`

**Comment by ilsaloving** [comment]
> We also discovered this falsepositive and wonder whats the problem aout whitelisting this specific hash. Disabling the whole check is not a trustworthy solution for such a new and promising project.

The problem is, we're not talking about just one rule.  There's tons.  The rootcheck rule database hasn't seen a modification in half a decade.  There needs to be a comprehensive review of the entire database.  sca isn't much better, with a ton of 'requirements' which are frequently debatable, and in some environments (eg: containers) don't make sense at all.

**Comment by opentokix** [comment]
Is Wazuh abandonware?

**Comment by ilsaloving** [comment]
Not even remotely.  There's very active development.  It's just that some of the checks wazuh does is based on sources that have not been well maintained.

**Comment by luclis** [comment]
For the rootkit detection, wazuh is based on ossec (and maybe for many over things). 
I patch the rootkit detection as said in [this issue](https://github.com/ossec/ossec-hids/issues/2020#issuecomment-1418331744), and everything works fine until last update.

I'm pretty sure they've change something there, but can't find it in github.

**Comment by khaisal13** [comment]
Still happen on wazuh v4.7.3, already change to 
diff        !bash|^/bin/sh|file\.h|proc\.h|^/bin/.*sh!

but the alert still show up

**Comment by KrisLowet** [comment]
Regarding https://github.com/ossec/ossec-hids/blob/master/src/rootcheck/db/rootkit_trojans.txt

This has to be:
`diff        !bash|^/bin/sh|file\.h|proc\.h|^/bin/.*sh!`

I replaced the whole file /var/ossec/etc/shared/default/rootkit_trojans.txt and that seems to be working.

**Comment by Elnadrion** [comment]
We also faced the issue recently

**Comment by SirStephanikus** [comment]
> > We also discovered this falsepositive and wonder whats the problem aout whitelisting this specific hash. Disabling the whole check is not a trustworthy solution for such a new and promising project.
> 
> The problem is, we're not talking about just one rule. There's tons. The rootcheck rule database hasn't seen a modification in half a decade. There needs to be a comprehensive review of the entire database. sca isn't much better, with a ton of 'requirements' which are frequently debatable, and in some environments (eg: containers) don't make sense at all.

SCA checks can only be edited correctly if someone has an active CIS Secureship Membership and access to CIS CAT Pro ... with customized profiles ... it makes sense!

**Comment by hulla-bulla** [comment]
Still happen for me with multiple fully new updated Ubuntu 22 and Wazuh 4.8.0

**Comment by ilsaloving** [comment]
Okay, so for SCA there's not much we can do.  Is there an override mechanism, similar to what you can do with alert rules?

**Comment by vkukk** [comment]
Still happens with 4.7.5.

**Comment by samaviya** [comment]
Still happening in 4.8

**Comment by aquagil** [comment]
Still happening in 4.8

**Comment by lucid8080** [comment]
The same issue on App version 4.8.0. I knew there was no way it could be right, because I did a fresh install of Linux for testing.

**Comment by iasdeoupxe** [comment]
@aquagil @lucid8080 IMHO no need to add additional comments like this, as long as this issue is open it probably won't be solved. And by not adding similar comments you will also avoid unnecessary notifications for users like me just monitoring this thread for a possible solution from the Wazuh team.

**Comment by tismofied** [comment]
Someone on google groups suggested to add the ignore function.
[link to possible fix](https://groups.google.com/g/wazuh/c/rN7VzfoYQ8w/m/hfFazCKfAAAJ)
I am hesitant to try it because I am not code savvy and afraid to break things.

**Comment by gand0rf** [comment]
The problem with the was the article points out that @tismofied linked, is you have to ignore the entire directory. I have this option setup on one of my agents to skip the docker folder on that agent. For this problem it could work, but is not the best option. The bin folder is where system wide executable on a linux system reside. You want to have eyes on it to be safe. Unless someone knows how to add a file ignore option instead of a directory ignore, another option needs to be found if a update to the db can not be done.

Also, Wazuh 4.8.1 Ubuntu 24.04.4 LTS same issue.

**Comment by vanyell** [comment]
Maybe this helps?
https://github.com/wazuh/wazuh/pull/3058

	<rootcheck>
		<ignore type="sregex">/usr/bin/diff</ignore>
	</rootcheck>

**Comment by vanyell** [comment]
rootcheck for /usr/bin/diff 
string check  'bash|^/bin/sh|file\.h|proc\.h|/dev/[^n]|^/bin/.*sh' (Generic).

running  `strings /usr/bin/diff` results in having _/dev/null, /dev/full_ in the list
```
c-maybe
clocale
/dev/null
/dev/full
mbuiter_multi_next
;*3$"
/usr/lib/debug/.dwz/x86_64-linux-gnu/diffutils.debug
8b2c724ea573d58b7e0fa35e0b41e03b896d27.debug
```

Maybe let's contribute to a Patreon account and get a CIS account to fix this mess? :) 

On the other hand... I thought SCA has replaced rootcheck ? why is rootcheck still in the mix?

**Comment by Serpensin** [comment]
Just installed on a fresh Ubuntu 24.04 and Wazuh 4.8.2 (no agent connected yet) and it also happened.

**Comment by SpecialEd792** [comment]
here to say the same

**Comment by iasdeoupxe** [comment]
New PR which could solve this: #25634

**Comment by marcinliszka12** [comment]
Wazuh version 4.9.0-1 same issue on Debian 12 fully patched.

**Comment by iasdeoupxe** [comment]
@marcinliszka12 No need for a "same issue":

https://github.com/wazuh/wazuh/issues/19346#issuecomment-2226930518

As long as #25634 isn't merged and this issue is closed the issue persists in all later releases.

**Comment by flygav1** [comment]
active as of 9/21/2024, new install and everything

**Comment by iasdeoupxe** [comment]
@flygav1 See previous comment(s).

@wazuh team: I would suggest locking this topic to avoid unnecessary comments and unnecessary notifications (caused by these) to people just having subscribed to the topic to get notified once it is solved.

The root cause has been identified, a pull request is open by a community member and any later comments won't add any valuable information to this issue in the current state.

**Comment by vikman90** [comment]
Hi all,

This issue has been fixed by @djosip:
- https://github.com/wazuh/wazuh/pull/25634

Best.

**Comment by street-grease-coder** [comment]
so what exactly is the idea here, we wait until there's an update for wazuh for this, and then the false-flags will be removed? Or will it be shown forever until we clean our logs? Do we just keep using it, and add exceptions, and learn about all false positives gradually, after freaking out?

**Comment by iasdeoupxe** [comment]
@street-grease-coder Fix seems to be planned for the next release 4.10.0:

![Screenshot_2024-10-20_17-40-05](https://github.com/user-attachments/assets/f984625a-2138-4706-817f-46aeba79c952)

**Comment by gand0rf** [comment]
> @street-grease-coder Fix seems to be planned for the next release 4.10.0:
> 
> ![Screenshot_2024-10-20_17-40-05](https://github.com/user-attachments/assets/f984625a-2138-4706-817f-46aeba79c952)
> 
> Existing fix had introduced some false negatives but for that new PR is open: #26160

Nice! Thanks for the update.

**Comment by iasdeoupxe** [comment]
> Same issue Ubuntu 22 Nov 13th 2024

See previous note / comment, they didn't released 4.10.0 yet AFAICT.

**Comment by tmcb1971** [comment]
And this is still going on here. As said previously on this thread, this kind of noise damages trust in the tool. But it isn't just this module which is broken, the listening ports check has the same effect. The major issue here is that Wazuh uses some really basic checks which can't handle the evolution of networking and file system operations. The analyst workflow has not been properly accounted for in the tool, so tuning has just turned into switch this rule on or off. We need much better workflow. It should be possible for rules to be fine tuned so that changing inodes or rotating IPv6 addresses can be accommodated without having to switch off potentially useful rules. This kind of noise will kill Wazuh if it doesn't get dealt with.

**Comment by yodaphone** [comment]
Dec 24 & this issue is still present

Ubuntu 22.04 with latest kernel

**Comment by iasdeoupxe** [comment]
@yodaphone https://github.com/wazuh/wazuh/issues/19346#issuecomment-2486743679 is still valid.

**Comment by JHugent** [comment]
Dec 30 & this issue is still present

Debian 12 with latest kernel

**Comment by iasdeoupxe** [comment]
@JHugent Have you noticed the previous comment? 🙂 

The issue will persist until the next Wazuh release and there is no need to add further comments here as the issue is known.

**Comment by iasdeoupxe** [comment]
Looks like 4.10.0 is out now including the fix from #25634 (Still not fully correct, see #26160 but better then nothing :-)):

https://github.com/wazuh/wazuh/blob/v4.10.0/ruleset/rootcheck/db/rootkit_trojans.txt#L43

**Comment by h1contr4st** [comment]
4.12.0 and still present for me.

**Comment by iasdeoupxe** [comment]
4.13.0 seems to contain a different fix: #28982

**Comment by dorancemc** [comment]
Wazuh v4.13.1
Debian 12
Same message:

```
full_log:Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\.h|proc\.h|/dev/[^n]|^/bin/.*sh' (Generic). timestamp:Oct 14, 2025 @ 05:14:33.118 
```

**Comment by iasdeoupxe** [comment]
@dorancemc Sure you're running Wazuh 4.13.1? Mine has below which largely differs from yours and it seems you are using outdated signatures.

```
# grep "diff" /var/ossec/etc/rootcheck/rootkit_trojans.txt
diff        !bash|^/bin/sh|file\.h|proc\.h|(/dev/($|[^nf]|n($|[^u])|nu($|[^l])|nul($|[^l])|null.+|f($|[^u])|fu($|[^l])|ful($|[^l])|full.+))|^/bin/.*sh!
```

Looks like you are using a highly outdated signature from even before the initial change in #26141

**Comment by andreasbrett** [comment]
> [@dorancemc](https://github.com/dorancemc) Sure you're running Wazuh 4.13.1? Mine has below which largely differs from yours and it seems you are using outdated signatures.
> 
> ```
> # grep "diff" /var/ossec/etc/rootcheck/rootkit_trojans.txt
> diff        !bash|^/bin/sh|file\.h|proc\.h|(/dev/($|[^nf]|n($|[^u])|nu($|[^l])|nul($|[^l])|null.+|f($|[^u])|fu($|[^l])|ful($|[^l])|full.+))|^/bin/.*sh!
> ```
> 
> Looks like you are using a highly outdated signature from even before the initial change in [#26141](https://github.com/wazuh/wazuh/pull/26141)

I have the same `full_log` message as @dorancemc but also the same signature in `/var/ossec/etc/rootcheck/rootkit_trojans.txt` that you posted. Weirdly I only get the event/alert for 1 of my servers - though my whole fleet is on the same Wazuh agent version (14.3.1).

**Comment by samuelfiery2** [comment]
> > [@dorancemc](https://github.com/dorancemc) Sure you're running Wazuh 4.13.1? Mine has below which largely differs from yours and it seems you are using outdated signatures.
> > ```
> > # grep "diff" /var/ossec/etc/rootcheck/rootkit_trojans.txt
> > diff        !bash|^/bin/sh|file\.h|proc\.h|(/dev/($|[^nf]|n($|[^u])|nu($|[^l])|nul($|[^l])|null.+|f($|[^u])|fu($|[^l])|ful($|[^l])|full.+))|^/bin/.*sh!
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Looks like you are using a highly outdated signature from even before the initial change in [#26141](https://github.com/wazuh/wazuh/pull/26141)
> 
> I have the same `full_log` message as [@dorancemc](https://github.com/dorancemc) but also the same signature in `/var/ossec/etc/rootcheck/rootkit_trojans.txt` that you posted. Weirdly I only get the event/alert for 1 of my servers - though my whole fleet is on the same Wazuh agent version (14.3.1).

Maybe that 1 server agent/var/ossec/etc/shared/rootkit_trojans.txt is not updated/overwritten with new config?

I checked mine and found these : 

The wazuh server have the newer signatures : 
`diff        !bash|^/bin/sh|file\.h|proc\.h|(/dev/($|[^nf]|n($|[^u])|nu($|[^l])|nul($|[^l])|null.+|f($|[^u])|fu($|[^l])|ful($|[^l])|full.+))|^/bin/.*sh!
`
On the agent however even after updated to 4.13.1,
```
grep diff /var/ossec/etc/shared/rootkit_trojans.txt
diff        !bash|^/bin/sh|file\.h|proc\.h|/dev/[^n]|^/bin/.*sh!
```

**Comment by iasdeoupxe** [comment]
> Maybe that 1 server agent/var/ossec/etc/shared/rootkit_trojans.txt is not updated/overwritten with new config?

If this is on a agent most likely one of these known issues reported to them (Wazuh team):
- #22425 
- #27234
- #21400

**Comment by p-romeo** [comment]
Still present 5/1/2025

**Comment by iasdeoupxe** [comment]
@p-romeo Really the same, or is this the new one known here:
- #32320
- #32142
- #35704

Because this here should be already fixed by them 4.13.1 released last year.
